standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -648,260 +431,85 @@ public OpenTxnsResponse openTxns(OpenTxnRequest rqst) throws MetaException {
+     * and opening T[4] parallel it is possible, that T[4] will be using the value from a snapshot with T[2,2],
+     * and we will have a lost update problem
+     */
+    acquireTxnLock(true);
+    // Measure the time from acquiring the sequence value, till committing in the TXNS table
+    StopWatch generateTransactionWatch = new StopWatch();
+    generateTransactionWatch.start();
 
-        LOG.debug("Going to commit");
-        dbConn.commit();
-        generateTransactionWatch.stop();
-        long elapsedMillis = generateTransactionWatch.getTime(TimeUnit.MILLISECONDS);
-        TxnType txnType = rqst.isSetTxn_type() ? rqst.getTxn_type() : TxnType.DEFAULT;
-        if (txnType != TxnType.READ_ONLY && elapsedMillis >= openTxnTimeOutMillis) {
-          /*
-           * The commit was too slow, we can not allow this to continue (except if it is read only,
-           * since that can not cause dirty reads).
-           * When calculating the snapshot for a given transaction, we look back for possible open transactions
-           * (that are not yet committed in the TXNS table), for TXN_OPENTXN_TIMEOUT period.
-           * We can not allow a write transaction, that was slower than TXN_OPENTXN_TIMEOUT to continue,
-           * because there can be other transactions running, that didn't considered this transactionId open,
-           * this could cause dirty reads.
-           */
-          LOG.error("OpenTxnTimeOut exceeded commit duration {}, deleting transactionIds: {}", elapsedMillis, txnIds);
-          deleteInvalidOpenTransactions(dbConn, txnIds);
-          dbConn.commit();
-          /*
-           * We do not throw RetryException directly, to not circumvent the max retry limit
-           */
-          throw new SQLException("OpenTxnTimeOut exceeded", MANUAL_RETRY);
-        }
-        return new OpenTxnsResponse(txnIds);
-      } catch (SQLException e) {
-        LOG.debug("Going to rollback: ", e);
-        rollbackDBConn(dbConn);
-        checkRetryable(e, "openTxns(" + rqst + ")");
-        throw new MetaException("Unable to select from transaction database " + StringUtils.stringifyException(e));
-      } finally {
-        close(null, stmt, dbConn);
-      }
-    } catch (RetryException e) {
-      return openTxns(rqst);
-    }
-  }
+    List<Long> txnIds = new OpenTxnsFunction(rqst, openTxnTimeOutMillis, transactionalListeners).execute(jdbcResource);
 
-  private List<Long> openTxns(Connection dbConn, OpenTxnRequest rqst)
-          throws SQLException, MetaException {
-    int numTxns = rqst.getNum_txns();
-    // Make sure the user has not requested an insane amount of txns.
-    int maxTxns = MetastoreConf.getIntVar(conf, ConfVars.TXN_MAX_OPEN_BATCH);
-    if (numTxns > maxTxns) {
-      numTxns = maxTxns;
-    }
-    List<PreparedStatement> insertPreparedStmts = null;
+    LOG.debug("Going to commit");
+    jdbcResource.getTransactionManager().getActiveTransaction().createSavepoint();
```

Comment:
why do we need this, it's not present in original code?

Reply:
It was there:

https://github.com/apache/hive/pull/4566/commits/f7974d8f6f365fe244ccc7e75da5f9ab5d85fc7a#diff-ccc50cf147261af64a10e20359f9c52f1d4797bc8e03e682d6a88891ee941f02L699

Reply:
what is the line number, is that `dbConn.commit();`? if yes, why don't we call `commit` instead of `getActiveTransaction().createSavepoint()`